## <algorithm>
1. **התחלת המשחק:**
   - הדפס הודעת פתיחה והסבר על הכללים.
     - דוגמה: "ברוכים הבאים ל-REVRSE! המטרה היא ללכוד כמה שיותר אסימוני יריב."
   - צור לוח משחק בגודל 8x8.
   - הנח ארבעה אסימונים במרכז הלוח (שניים שחורים ושניים לבנים באלכסון).
   - קבע שחקן ראשון (לדוגמה, שחור).
2. **לולאת משחק עיקרית:**
   - כל עוד המשחק לא נגמר (הלוח לא מלא או שאף שחקן לא יכול לבצע מהלך):
     - תור השחקן הנוכחי:
       - בקש מהשחקן להזין את הקואורדינטות של המהלך (לדוגמה, D3).
       - בדוק אם המהלך חוקי:
         - האסימון חייב להיות מונח על משבצת ריקה.
         - האסימון חייב ללכוד לפחות אסימון אחד של היריב בין האסימון שלו לאסימון אחר שלו.
         - דוגמה: אם יש רצף "שחור-לבן-לבן-שחור" ושחור מונח, יש להפוך את הלבנים לשחורים.
       - אם המהלך חוקי:
         - הנח את האסימון של השחקן על הלוח.
         - הפוך את האסימונים של היריב שנתפסו.
         - הצג את מצב הלוח הנוכחי.
         - העבר את התור לשחקן הבא.
       - אם המהלך לא חוקי:
         - הדפס הודעת שגיאה ובקש מהשחקן להזין מהלך אחר.
3. **סיום המשחק:**
   - המשחק מסתיים כאשר הלוח מלא או כאשר אף שחקן לא יכול לבצע מהלך חוקי.
   - ספור את מספר האסימונים של כל שחקן.
   - הכרז על המנצח (השחקן עם הכי הרבה אסימונים).
   - דוגמה: "המשחק נגמר! שחור: 32, לבן: 30. שחור ניצח!"
   - שאל את השחקנים אם הם רוצים לשחק שוב. אם כן, אתחל משחק חדש.

## <mermaid>
```mermaid
flowchart TD
    Start[התחל משחק] --> Init[אתחול המשחק];
    Init --> Welcome[הדפס הודעת פתיחה והסבר על הכללים];
    Welcome --> CreateBoard[צור לוח 8x8];
    CreateBoard --> PlaceInitialPieces[הנח ארבעה אסימונים במרכז];
    PlaceInitialPieces --> SetFirstPlayer[קבע שחקן ראשון (שחור)];
    SetFirstPlayer --> GameLoop[לולאת המשחק];
    GameLoop -- המשחק לא נגמר --> CurrentPlayerTurn[תור השחקן הנוכחי];
    CurrentPlayerTurn --> GetMove[קבל קואורדינטות מהשחקן];
    GetMove --> IsValidMove[בדוק אם המהלך חוקי];
    IsValidMove -- מהלך חוקי --> PlacePiece[הנח את האסימון על הלוח];
    PlacePiece --> FlipOpponentPieces[הפוך אסימונים של היריב שנלכדו];
    FlipOpponentPieces --> DisplayBoard[הצג את מצב הלוח];
    DisplayBoard --> SwitchPlayer[העבר את התור לשחקן הבא];
    SwitchPlayer --> GameLoop;
    IsValidMove -- מהלך לא חוקי --> InvalidMoveMessage[הדפס הודעת שגיאה];
    InvalidMoveMessage --> GetMove;
    GameLoop -- המשחק נגמר --> EndGame[סיום המשחק];
    EndGame --> CountPieces[ספור אסימונים של כל שחקן];
    CountPieces --> DeclareWinner[הכרז על המנצח];
    DeclareWinner --> PlayAgain[שאל אם רוצים לשחק שוב];
    PlayAgain -- כן --> Init;
    PlayAgain -- לא --> End[סיום];
```

## <explanation>
- **ייבואים (Imports)**:
   - לא צוינו ייבואים בקוד זה. המשחק מתוכנן להיות ממומש בפייתון, ללא תלות בחבילות חיצוניות.

- **מחלקות (Classes)**:
   - לא נעשה שימוש במחלקות בקוד זה. המשחק מתוכנן להיות ממומש באמצעות פונקציות ומשתנים.

- **פונקציות (Functions)**:
  - הפונקציות הדרושות למימוש המשחק יהיו:
    - `init_game()`: מאתחלת את המשחק, יוצרת את הלוח, מניחה את האסימונים ההתחלתיים וקובעת שחקן ראשון.
    - `get_player_move()`: מקבלת קואורדינטות מהשחקן ומוודאת את תקינות הקלט.
      - פרמטרים: אין.
      - מחזירה: קואורדינטות של המהלך (לדוגמה, D3) או שגיאה אם הקלט לא תקין.
    - `is_valid_move(board, row, col, player)`: בודקת אם המהלך חוקי.
      - פרמטרים: הלוח, שורה, עמודה, השחקן הנוכחי.
      - מחזירה: `True` אם המהלך חוקי, `False` אם לא.
    - `place_piece(board, row, col, player)`: מניחה אסימון על הלוח והופכת אסימונים של היריב שנתפסו.
      - פרמטרים: הלוח, שורה, עמודה, השחקן הנוכחי.
      - מחזירה: הלוח המעודכן.
    - `display_board(board)`: מציגה את מצב הלוח.
      - פרמטרים: הלוח.
      - מחזירה: אין, מדפיסה את הלוח לקונסולה.
    - `switch_player(current_player)`: מעבירה את התור לשחקן הבא.
      - פרמטרים: השחקן הנוכחי.
      - מחזירה: השחקן הבא.
    - `check_game_over(board)`: בודקת אם המשחק נגמר (הלוח מלא או שאף שחקן לא יכול לבצע מהלך).
       - פרמטרים: הלוח.
       - מחזירה: `True` אם המשחק נגמר, `False` אם לא.
    - `count_pieces(board)`: סופרת את האסימונים של כל שחקן.
      - פרמטרים: הלוח.
      - מחזירה: מספר האסימונים של כל שחקן (שחור ולבן).
    - `declare_winner(black_count, white_count)`: מכריזה על המנצח.
       - פרמטרים: מספר האסימונים של שחור, מספר האסימונים של לבן.
       - מחזירה: אין, מדפיסה את המנצח לקונסולה.

- **משתנים (Variables)**:
  - `board`: רשימה דו-ממדית (list) המייצגת את לוח המשחק (8x8), בו כל תא יכול להיות ריק (" "), אסימון שחור ("B") או אסימון לבן ("W").
  - `current_player`: מציין את השחקן הנוכחי ("B" או "W").

- **בעיות אפשריות או תחומים לשיפור**:
  - אימות קלט המשתמש: יש צורך לוודא שהקלט של המשתמש (הקואורדינטות של המהלך) תקין, ולטפל במקרים בהם הקלט לא תקין.
  - הגיון המהלך החוקי: צריך לממש בצורה מדויקת את הגיון בדיקת המהלך החוקי של המשחק.
  - ממשק משתמש: אין גרפיקה, רק ממשק טקסטואלי. ניתן לשפר על ידי הוספת ממשק משתמש גרפי.
  - אפשרות משחק נגד המחשב: אין אפשרות לשחק נגד מחשב.

- **קשרים עם חלקים אחרים בפרויקט**:
  - קוד זה יכול לשמש כחלק ממשחק רחב יותר, או כמודול עצמאי.
  - הוא לא תלוי באופן ספציפי בחלקים אחרים של הפרויקט, אבל יכול להשתמש בשירותים משותפים (לדוגמה, ממשק משתמש גרפי).