## <algorithm>

1. **אתחול המשחק:**
   - הצגת הודעת פתיחה והסבר על חוקי המשחק.
     - דוגמה: "ברוכים הבאים ל-ROCKSP! המטרה היא לחסום את היריב על ידי הצבת אבנים על הלוח."
   - יצירת לוח משחק בגודל 6x6.
     - דוגמה: יצירת מערך דו-ממדי המייצג את הלוח, כאשר כל תא ריק בתחילה.
   - התחלת המשחק עם השחקן הראשון.
2. **מהלך השחקן:**
   - השחקן בוחר תא להצבת אבן על ידי הזנת קואורדינטות.
     - דוגמה: השחקן מזין "B2".
   - בדיקת תקינות המהלך:
     - האם התא פנוי?
       - דוגמה: אם התא B2 כבר מכיל אבן, המהלך אינו תקין.
     - האם הקואורדינטות בתוך גבולות הלוח?
       - דוגמה: אם השחקן מזין "G7", המהלך אינו תקין.
   - אם המהלך תקין:
     - הצבת אבן של השחקן בתא הנבחר.
       - דוגמה: הצבת "R" (אבן של השחקן) בתא B2.
     - הצגת הלוח המעודכן.
     - מעבר לתור של השחקן הבא.
   - אם המהלך אינו תקין:
     - הצגת הודעת שגיאה.
       - דוגמה: "מהלך לא תקין, נסה שוב".
     - בקשה מהשחקן לבצע מהלך נוסף.
3. **בדיקת תנאי ניצחון:**
   - לאחר כל מהלך, בדיקה האם היריב יכול לבצע מהלך נוסף.
     - דוגמה: מעבר על כל התאים הריקים ובדיקה האם ניתן להציב שם אבן.
   - אם היריב לא יכול לבצע מהלך:
     - הכרזה על המנצח.
       - דוגמה: "המשחק הסתיים! שחקן 1 ניצח."
4. **בדיקת תנאי סיום משחק:**
   - האם הגיעו למספר המהלכים המקסימלי (לדוגמה, 20)?
   - אם כן:
     - הכרזה על תיקו.
       - דוגמה: "המשחק הסתיים! תיקו."
5. **סיום המשחק:**
   - שאלה האם לשחק שוב.
   - אם כן, אתחול המשחק מחדש.
   - אם לא, סיום התוכנית.

## <mermaid>
```mermaid
flowchart TD
    Start[התחלת המשחק] --> InitializeGame[אתחול המשחק: <br>  - הצגת הודעה והסבר על החוקים<br>  - יצירת לוח משחק בגודל 6x6<br>  - התחלת משחק]
    InitializeGame --> PlayerTurn[תור השחקן]
    PlayerTurn --> GetMove[קבלת מהלך מהשחקן]
    GetMove --> ValidateMove{בדיקת תקינות המהלך?}
    ValidateMove -- מהלך לא תקין --> InvalidMove[הודעת שגיאה: <br> "מהלך לא תקין, נסה שוב."]
    InvalidMove --> PlayerTurn
    ValidateMove -- מהלך תקין --> PlaceStone[הצבת אבן על הלוח]
    PlaceStone --> DisplayBoard[הצגת הלוח המעודכן]
    DisplayBoard --> CheckWinCondition[בדיקת תנאי ניצחון: <br> האם היריב יכול לבצע מהלך?]
    CheckWinCondition -- יריב לא יכול לבצע מהלך --> PlayerWins[הכרזת ניצחון]
    PlayerWins --> EndGame[סיום המשחק]
    CheckWinCondition -- יריב יכול לבצע מהלך --> CheckMaxMoves[בדיקת תנאי סיום: <br> האם הגיעו למקסימום מהלכים?]
    CheckMaxMoves -- הגעה למקסימום מהלכים --> TieGame[הכרזה על תיקו]
    TieGame --> EndGame
    CheckMaxMoves -- לא הגעה למקסימום מהלכים --> NextPlayerTurn[מעבר לתור השחקן הבא]
    NextPlayerTurn --> PlayerTurn
    EndGame --> PlayAgain{לשחק שוב?}
    PlayAgain -- כן --> Start
    PlayAgain -- לא --> Stop[סיום התוכנית]
```

**ניתוח התלויות:**
אין תלויות חיצוניות. כל הפונקציות ומשתנים מתבצעים בתוך המשחק עצמו.

## <explanation>

**ייבוא (Imports):**
אין ייבוא בתיאור המשחק. כל הלוגיקה מתבצעת בתוך הקוד עצמו.
**מחלוקות (Classes):**
לא נעשה שימוש במחלקות בתיאור המשחק. המשחק פשוט ומסתמך על פונקציות ולולאות.

**פונקציות (Functions):**
בתיאור המשחק אין פונקציות מפורשות, אבל אפשר להגדיר כמה פונקציות עיקריות:

1.  **`init_game()`**:
    -   **פרמטרים**: אין.
    -   **ערך מוחזר**: לוח משחק התחלתי (רשימה דו-ממדית).
    -   **מטרה**: לאתחל את המשחק, ליצור את הלוח ולהציג את הודעת הפתיחה.
    -   **דוגמא**: `board = init_game()`

2.  **`get_player_move()`**:
    -   **פרמטרים**: שחקן נוכחי (מספר).
    -   **ערך מוחזר**: קואורדינטות של המהלך (מחרוזת).
    -   **מטרה**: לקבל את המהלך מהשחקן, ולבדוק שזהו מהלך חוקי.
    -   **דוגמא**: `move = get_player_move(1)`

3.  **`validate_move(board, move)`**:
    -   **פרמטרים**: לוח משחק, מהלך שחקן (קואורדינטות).
    -   **ערך מוחזר**: True אם המהלך תקין, False אם לא.
    -   **מטרה**: לבדוק שהמהלך תקין (התא פנוי ובתוך גבולות הלוח).
    -   **דוגמא**: `is_valid = validate_move(board, "B2")`

4.  **`place_stone(board, move, player)`**:
    -   **פרמטרים**: לוח משחק, מהלך שחקן, שחקן נוכחי.
    -   **ערך מוחזר**: לוח מעודכן.
    -   **מטרה**: להציב את האבן של השחקן בתא הנבחר.
    -   **דוגמא**: `updated_board = place_stone(board, "B2", 1)`

5.  **`check_win_condition(board, current_player)`**:
    -   **פרמטרים**: לוח משחק, שחקן נוכחי.
    -   **ערך מוחזר**: True אם השחקן השני לא יכול לבצע מהלך, False אם כן.
    -   **מטרה**: לבדוק אם היריב לא יכול לבצע מהלך.
    -   **דוגמא**: `is_win = check_win_condition(board, 1)`

6. **`check_max_moves(move_count)`**:
    -   **פרמטרים**: מספר המהלכים שנעשו.
    -   **ערך מוחזר**: True אם הגענו למקסימום מהלכים, False אם לא.
    -   **מטרה**: לבדוק האם הגענו למקסימום המהלכים המוגדרים למשחק.
    -   **דוגמא**: `is_max = check_max_moves(move_count)`
7.  **`display_board(board)`**:
    -   **פרמטרים**: לוח משחק.
    -   **ערך מוחזר**: אין.
    -   **מטרה**: להציג את הלוח למשתמש.
    -   **דוגמא**: `display_board(board)`

**משתנים (Variables):**

*   `board`: רשימה דו-ממדית המייצגת את לוח המשחק. כל תא יכול להיות ריק או מכיל אבן של שחקן.
*   `current_player`: מספר השחקן הנוכחי (1 או 2).
*   `move`: קואורדינטות המהלך שבוצע על ידי השחקן.
*  `move_count`: מספר המהלכים שבוצעו במשחק.
*  `max_moves`: מספר המהלכים המקסימלי המוגדר למשחק.

**בעיות אפשריות ותחומים לשיפור:**

*   **בדיקת קלט**: צריך לוודא שהשחקן מכניס קלט תקין (פורמט נכון של קואורדינטות, ולא קלט שאינו מוגדר).
*   **מימוש בדיקת ניצחון**: צריך לממש את הלוגיקה לבדיקה האם היריב לא יכול לבצע מהלך.
*   **משחק נגד מחשב**: כרגע זה משחק רק בין שני שחקנים. צריך להוסיף אופציה לשחק נגד מחשב.
*   **ממשק גרפי**: הוספת ממשק גרפי יעזור להפוך את המשחק לידידותי יותר למשתמש.
*   **גודל לוח**: אפשרות לבחור גודל לוח משחק שונה, כמו 8x8 או 10x10.

**שרשרת קשרים עם חלקים אחרים בפרויקט:**
המשחק עומד בפני עצמו ואין לו תלות ישירה בחלקים אחרים בפרויקט.